File: sdk/margined/_extract.py
"""
Usage extraction from provider response objects.

Handles Anthropic Messages, OpenAI Chat Completions / Responses, and any
object exposing a compatible `.usage`. Cache tokens are split out so cost
reflects provider cache discounts:

  Anthropic usage:  input_tokens (uncached), output_tokens,
                    cache_read_input_tokens, cache_creation_input_tokens
  OpenAI usage:     prompt_tokens (INCLUDES cached), completion_tokens,
                    prompt_tokens_details.cached_tokens
"""

from __future__ import annotations

import time
import uuid
from typing import Any, Optional

from ._pricing import compute_cost, detect_provider


def _get(obj: Any, name: str) -> int:
    value = getattr(obj, name, None)
    if value is None and isinstance(obj, dict):
        value = obj.get(name)
    try:
        return int(value) if value is not None else 0
    except (TypeError, ValueError):
        return 0


def extract_usage(response: Any) -> Optional[dict[str, int]]:
    """Return normalized token counts from a provider response, or None."""
    usage = getattr(response, "usage", None)
    if usage is None and isinstance(response, dict):
        usage = response.get("usage")
    if usage is None:
        return None
    return usage_from_obj(usage)


def usage_from_obj(usage: Any) -> dict[str, int]:
    # Anthropic-style
    input_tokens = _get(usage, "input_tokens")
    output_tokens = _get(usage, "output_tokens")
    cache_read = _get(usage, "cache_read_input_tokens")
    cache_write = _get(usage, "cache_creation_input_tokens")

    if input_tokens == 0 and output_tokens == 0:
        # OpenAI-style
        prompt = _get(usage, "prompt_tokens")
        completion = _get(usage, "completion_tokens")
        details = getattr(usage, "prompt_tokens_details", None)
        if details is None and isinstance(usage, dict):
            details = usage.get("prompt_tokens_details")
        cached = _get(details, "cached_tokens") if details is not None else 0
        input_tokens = max(prompt - cached, 0)  # uncached portion
        output_tokens = completion
        cache_read = cached
        cache_write = 0

    return {
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "cache_read_tokens": cache_read,
        "cache_write_tokens": cache_write,
    }


def build_event(
    response: Any,
    user_id: str,
    feature: str,
    run_id: Optional[str] = None,
    metadata: Optional[dict[str, Any]] = None,
    model: Optional[str] = None,
) -> Optional[dict[str, Any]]:
    """Build an ingest event from a provider response. None if no usage found."""
    usage = extract_usage(response)
    if usage is None:
        return None
    resolved_model = model or getattr(response, "model", None) or "unknown"
    return build_event_from_usage(usage, resolved_model, user_id, feature, run_id, metadata)


def build_event_from_usage(
    usage: dict[str, int],
    model: str,
    user_id: str,
    feature: str,
    run_id: Optional[str] = None,
    metadata: Optional[dict[str, Any]] = None,
) -> dict[str, Any]:
    cost = compute_cost(
        model,
        usage["input_tokens"],
        usage["output_tokens"],
        usage["cache_read_tokens"],
        usage["cache_write_tokens"],
    )
    return {
        # Idempotency key: the ingest API dedupes on (project, event_id), so a
        # retried flush can never double-count cost.
        "event_id": str(uuid.uuid4()),
        "customer_id": str(user_id),
        "feature": str(feature),
        "run_id": run_id,
        "model": model,
        "provider": detect_provider(model),
        "input_tokens": usage["input_tokens"],
        "output_tokens": usage["output_tokens"],
        "cache_read_tokens": usage["cache_read_tokens"],
        "cache_write_tokens": usage["cache_write_tokens"],
        "cost_usd": cost,
        "occurred_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "metadata": metadata or {},
    }

File: sdk-node/src/extract.ts
/**
 * Usage extraction from provider responses (Anthropic Messages, OpenAI Chat
 * Completions) and event construction. Cache tokens are split out; OpenAI's
 * prompt_tokens INCLUDES cached tokens, so the uncached remainder is stored.
 */

import { computeCost, detectProvider } from "./pricing.js";

export interface Usage {
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens: number;
  cacheWriteTokens: number;
}

export interface MarginedEvent {
  event_id: string;
  customer_id: string;
  feature: string;
  run_id: string | null;
  model: string;
  provider: string;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
  cost_usd: number;
  occurred_at: string;
  metadata: Record<string, unknown>;
}

function num(obj: unknown, key: string): number {
  if (obj == null || typeof obj !== "object") return 0;
  const value = (obj as Record<string, unknown>)[key];
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

export function usageFromObject(usage: unknown): Usage {
  // Anthropic-style
  let inputTokens = num(usage, "input_tokens");
  let outputTokens = num(usage, "output_tokens");
  let cacheRead = num(usage, "cache_read_input_tokens");
  let cacheWrite = num(usage, "cache_creation_input_tokens");

  if (inputTokens === 0 && outputTokens === 0) {
    // OpenAI-style
    const prompt = num(usage, "prompt_tokens");
    const completion = num(usage, "completion_tokens");
    const details = (usage as Record<string, unknown> | null)?.["prompt_tokens_details"];
    const cached = num(details, "cached_tokens");
    inputTokens = Math.max(prompt - cached, 0);
    outputTokens = completion;
    cacheRead = cached;
    cacheWrite = 0;
  }

  return { inputTokens, outputTokens, cacheReadTokens: cacheRead, cacheWriteTokens: cacheWrite };
}

export function extractUsage(response: unknown): Usage | null {
  if (response == null || typeof response !== "object") return null;
  const usage = (response as Record<string, unknown>).usage;
  if (usage == null) return null;
  return usageFromObject(usage);
}

export interface Tags {
  userId: string;
  feature: string;
  runId?: string | null;
  metadata?: Record<string, unknown>;
}

export function buildEventFromUsage(usage: Usage, model: string, tags: Tags): MarginedEvent {
  return {
    // Idempotency key — the ingest API dedupes on it, so retried flushes
    // can never double-count cost.
    event_id: crypto.randomUUID(),
    customer_id: String(tags.userId),
    feature: String(tags.feature),
    run_id: tags.runId ?? null,
    model,
    provider: detectProvider(model),
    input_tokens: usage.inputTokens,
    output_tokens: usage.outputTokens,
    cache_read_tokens: usage.cacheReadTokens,
    cache_write_tokens: usage.cacheWriteTokens,
    cost_usd: computeCost(
      model,
      usage.inputTokens,
      usage.outputTokens,
      usage.cacheReadTokens,
      usage.cacheWriteTokens,
    ),
    occurred_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    metadata: tags.metadata ?? {},
  };
}

export function buildEvent(response: unknown, tags: Tags): MarginedEvent | null {
  const usage = extractUsage(response);
  if (usage == null) return null;
  const model =
    (response as Record<string, unknown> | null)?.["model"] != null
      ? String((response as Record<string, unknown>)["model"])
      : "unknown";
  return buildEventFromUsage(usage, model, tags);
}
